internal/pkg/service/common/etcdop/op/atomic_core.go

```diff
@@ -2,112 +2,217 @@ package op
+	// we can return operations without wrapping them to a TXN.
+	if v.processorFactory == nil && len(v.locks) == 0 {
+		return slices.Clone(v.writePhase)
 	}
+
+	return []HighLevelFactory{func(ctx context.Context) Op {
+		txn, err := v.writeTxn(ctx)
+		if err == nil {
+			return txn
+		} else {
+			return ErrorOp(err)
+		}
+	}}
+}
+
+// Core - to match AtomicOpInterface.
+func (v *AtomicOpCore) Core() *AtomicOpCore {
 	return v
 }
 
-func (v *AtomicOpCore) OnReadOrErr(fns ...func(ctx context.Context) error) *AtomicOpCore {
-	for _, fn := range fns {
-		v.ReadOrErr(func(ctx context.Context) (Op, error) {
-			return nil, fn(ctx)
-		})
+func (v *AtomicOpCore) Empty() bool {
+	return len(v.locks) == 0 && len(v.readPhase) == 0 && len(v.writePhase) == 0
+}
+
+// AddFrom merges operations from some other atomic operation.
+func (v *AtomicOpCore) AddFrom(ops ...AtomicOpInterface) *AtomicOpCore {
+	for _, op := range ops {
+		v.readPhase = append(v.readPhase, op.ReadPhaseOps()...)
+		v.writePhase = append(v.writePhase, op.WritePhaseOps()...)
 	}
 	return v
 }
 
-func (v *AtomicOpCore) ReadOrErr(factories ...HighLevelFactory) *AtomicOpCore {
-	v.readPhase = append(v.readPhase, factories...)
+func (v *AtomicOpCore) RequireLock(lock Mutex) *AtomicOpCore {
+	v.locks = append(v.locks, lock)
 	return v
 }
 
-func (v *AtomicOpCore) Write(factories ...func(ctx context.Context) Op) *AtomicOpCore {
-	for _, fn := range factories {
-		v.WriteOrErr(func(ctx context.Context) (Op, error) {
-			return fn(ctx), nil
-		})
-	}
-	return v
+func (v *AtomicOpCore) newEmpty() *AtomicOpCore {
+	out := newAtomicCore(v.client)
+	out.checkPrefixKey = v.checkPrefixKey
+	return out
+}
+
+func (v *AtomicOpCore) setProcessorFactory(fn atomicOpCoreProcessorFactory) {
+	v.processorFactory = fn
 }
 
-func (v *AtomicOpCore) BeforeWrite(fns ...func(ctx context.Context)) *AtomicOpCore {
-	for _, fn := range fns {
-		v.WriteOrErr(func(ctx context.Context) (Op, error) {
-			fn(ctx)
-			return nil, nil
-		})
+func (v *AtomicOpCore) locksIfs() (cmps []etcd.Cmp, err error) {
+	for _, lock := range v.locks {
+		if key := lock.Key(); key == "" || key == "\x00" {
+			return nil, NotLockedError{}
+		}
+		cmps = append(cmps, lock.IsOwner())
 	}
-	return v
+	return cmps, nil
 }
 
-func (v *AtomicOpCore) BeforeWriteOrErr(fns ...func(ctx context.Context) error) *AtomicOpCore {
-	for _, fn := range fns {
-		v.WriteOrErr(func(ctx context.Context) (Op, error) {
-			return nil, fn(ctx)
-		})
+func (v *AtomicOpCore) readTxn(ctx context.Context, tracker *TrackerKV) (*TxnOp[NoResult], error) {
+	// Create READ transaction
+	readTxn := Txn(tracker)
+	for _, opFactory := range v.readPhase {
+		op := opFactory(ctx)
+		if errOp, ok := op.(*errorOp); ok {
+			return nil, errOp.err
+		} else if op != nil {
+			readTxn.Merge(op)
+		}
 	}
-	return v
+
+	// Stop the READ phase, if a lock is not locked
+	if lockIfs, err := v.locksIfs(); err == nil && len(lockIfs) > 0 {
+		readTxn.Merge(Txn(nil).
```

Comment:
Why this `Txn` is without `tracker`? The actual READ phase does not require it, only the transaction?

Reply:
`readTxn` is using `tracker`:
https://github.com/keboola/keboola-as-code/blob/7a99f5a9d36728982b730f2a90979a710c3caa86/internal/pkg/service/common/etcdop/op/atomic_core.go#L134

Here is `Txn(client=nil)`, because the transaction is not invoked directly, but is merged to the `readTxn`.

-----------

We have two options, how it may be implemented:

1. `op(client).Do(ctx)` - the etcd client is set on operation creation
 - `return op(client)`
 - `repository.SomeOp().Do(ctx)`
2. `op().Do(ctx, client)` - the etcd client is passed on invocation
 - `return op()`
 - `repository.SomeOp().Do(ctx, client)`
-------------

I selected the option `1.`, because 
- the (returned) operation can be always invoked in some other part of the service, 
- and no additional `client` dependency is needed
- OR operation can be merged with some other operation

cc @jachym-tousek-keboola @hosekpeter